File: dotnet/Metrics/GlobalRatingsGivenByEachUserPerCategory.cs
namespace Southworks.Demos.Spark.Metrics
{
    using Microsoft.Spark.Sql;
    using static Microsoft.Spark.Sql.Functions;

    public class GlobalRatingsGivenByEachUserPerCategory : IMetricsSolver
    {
        public DataFrame Run(DataFrame moviesDataFrame, DataFrame ratingsDataFrame)
        {
            // Find the users that rated at least 250 movies.
            var qualifyingUsers = ratingsDataFrame
                .GroupBy("userId")
                .Count()
                .Filter("count >= 250")
                .WithColumnRenamed("userId", "r_userId");

            // Find all the movies rated by qualifying users.
            var moviesAndQualifyingUsersRating = ratingsDataFrame
                .Join(qualifyingUsers, ratingsDataFrame.Col("userId").EqualTo(qualifyingUsers.Col("r_userId")))
                .Select("userId", "movieId", "rating")
                .WithColumnRenamed("movieId", "r_movieId");

            // Explode categories into a row per movie - category combination.
            // Group the categories by userId and get each average rating.
            var explodedCategoriesAggregatedByAverageRatingPerUser = moviesDataFrame
                .Join(moviesAndQualifyingUsersRating,
                    moviesDataFrame.Col("movieId").EqualTo(moviesAndQualifyingUsersRating.Col("r_movieId")))
                .WithColumn("genres", Split(Col("genres"), "\\|"))
                .Select(Col("userId"),
                    Col("movieId"),
                    Col("rating"),
                    Explode(Col("genres")))
                .WithColumnRenamed("col", "genre")
                .GroupBy("userId", "genre")
                .Agg(Avg("rating"))
                .OrderBy(Col("userId").Asc())
                .Select("userId", "avg(rating)", "genre")
                .WithColumnRenamed("avg(rating)", "average score")
                .Limit(20);

            return explodedCategoriesAggregatedByAverageRatingPerUser;
        }
    }
}

File: scala/src/main/scala/com/southworks/demos/metrics/implementation/GlobalRatingsGivenByEachUserPerCategory.scala
package com.southworks.demos.metrics.implementation

import com.southworks.demos.metrics.MetricsSolver
import org.apache.spark.sql.functions.{avg, col, explode, split}
import org.apache.spark.sql.{Dataset, Row}

class GlobalRatingsGivenByEachUserPerCategory extends MetricsSolver {
  override def run(moviesDataset: Dataset[Row], ratingsDataset: Dataset[Row]): Dataset[Row] = {
    // Find the users that rated at least 250 movies.
    val qualifyingUsers = ratingsDataset
      .groupBy("userId")
      .count
      .filter("count >= 250")
      .withColumnRenamed("userId", "r_userId")

    // Find all the movies rated by qualifying users.
    val moviesAndQualifyingUsersRating = ratingsDataset
      .join(qualifyingUsers, ratingsDataset.col("userId").equalTo(qualifyingUsers.col("r_userId")))
      .select("userId", "movieId", "rating")
      .withColumnRenamed("movieId", "r_movieId")

    // Explode categories into a row per movie - category combination.
    // Group the categories by userId and get each average rating.
    val explodedCategoriesAggregatedByAverageRatingPerUser = moviesDataset
      .join(moviesAndQualifyingUsersRating, moviesDataset.col("movieId").equalTo(moviesAndQualifyingUsersRating.col("r_movieId")))
      .withColumn("genres", split(col("genres"), "\\|"))
      .select(col("userId"), col("movieId"), col("rating"), explode(col("genres")))
      .withColumnRenamed("col", "genre")
      .groupBy("userId", "genre")
      .agg(avg("rating"))
      .orderBy(col("userId").asc)
      .select("userId", "avg(rating)", "genre")
      .withColumnRenamed("avg(rating)", "average score")
      .limit(20)

    explodedCategoriesAggregatedByAverageRatingPerUser
  }
}
